pkg/virt-launcher/virtwrap/access-credentials/access_credentials.go

```diff
@@ -285,16 +285,15 @@ func (l *AccessCredentialManager) agentSetFilePermissions(domName string, filePa
 }
 
 func (l *AccessCredentialManager) agentSetUserPassword(domName string, user string, password string) error {
-
-	base64Str := base64.StdEncoding.EncodeToString([]byte(password))
-
-	cmdSetPassword := fmt.Sprintf(`{"execute":"guest-set-user-password", "arguments": {"username":"%s", "password": "%s", "crypted": false }}`, user, base64Str)
-
-	_, err := l.virConn.QemuAgentCommand(cmdSetPassword, domName)
+	domain, err := l.virConn.LookupDomainByName(domName)
 	if err != nil {
+		log.Log.Errorf("Domain lookup failed: %v", err)
 		return err
 	}
-	return nil
+	defer domain.Free()
+
+	encodedPassword := base64.StdEncoding.EncodeToString([]byte(password))
+	return domain.SetUserPassword(user, encodedPassword, 0)
```

Comment:
Are you sure the password still needs to be encoded?

It looks like libvirt handles this for us eventually:

https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainSetUserPassword

https://github.com/libvirt/libvirt/blob/7f07cf5a72434271b78d6b25a8132f0436e28f52/src/qemu/qemu_agent.c#L2152-L2176

Reply:
At least, the flag should be 1, no?

Reply:
No as that would suggest the password had been encrypted. All passwords need to be base64 encoded before being passed to QEMU regardless of also being encrypted:

https://github.com/qemu/qemu/blob/56c6e249b6988c1b6edc2dd34ebb0f1e570a1365/qga/qapi-schema.json#L1095-L1121

Reply:
Had to check the actual implementation within the guest agent but on linux it's just using `chpasswd` that already encrypts the passwords via PAM by default so again additional encryption shouldn't be required:

https://github.com/qemu/qemu/blob/56c6e249b6988c1b6edc2dd34ebb0f1e570a1365/qga/commands-posix.c#L835-L881

Reply:
:man_facepalming:  encode vs encrypt :man_facepalming: